### DISCUSSION TITLE
How to specify build-time configuration variables

### DISCUSSION THREAD

**Discussion by jrr** [remix-run/remix#8924] [Q&A]
The [Environment Variables](https://remix.run/docs/en/main/guides/envvars#environment-variables) doc describes how env vars are meant for runtime, and contrasts this with "build arguments".

I'm totally onboard with this, but as far as I can tell "build arguments" aren't a thing in remix. Is there a way I can define a variable to be consumed at build time? I'm looking for something analogous to [REACT_APP_](https://create-react-app.dev/docs/adding-custom-environment-variables/) variables, whether expressed via env vars or via some other mechanism.

**Comment by gustavopch** [comment]
If you're using Vite, you can inject env vars as described here: https://vitejs.dev/guide/env-and-mode

**Reply by jrr** [reply]
That looks great! Unfortunately not using Vite yet, so I need a solution that works with the old compiler. (though I look forward to migrating to the Vite solution when I can!)

**Reply by gustavopch** [reply]
I believe you'll need to write a script to go through the build files and replace each variable with the value you want. Or... just get the values from your server. I was convinced by Ryan that that's the way to go.

**Reply by sergiodxa** [reply]
@jrr old compiler there's no way, but you could add a prebuild script that generates a file with your configuration variables and then import in, so when you run `remix build` it will already have imported the file with the correct values

**Reply by jrr** [reply]
Yeah, that's what I wound up doing, adding a code generator step and a turbo dependency. It's just a bummer that I can only have one version on the filesystem at a time, whereas I could e.g. have `FOO=BAR remix dev` and `FOO=BAZ remix dev` running simultaneously if it worked like env vars.
